File: frontend_v2/src/features/financial/lib/indirect-review.ts
import { correctionMinor } from "./correction-contract"
import { z } from "zod"
const minor = z.string().regex(/^(0|-?[1-9][0-9]*)$/)
const methodId = z.enum(["net_worth", "bank_deposits", "expenditure", "cash_t"])
export const indirectCatalog = z.object({
  case_id: z.string().uuid(),
  reference: z.string().url(),
  methods: z.array(
    z.object({
      id: methodId,
      label: z.string(),
      reference_section: z.string(),
      terms: z.array(
        z.object({
          id: z.string(),
          label: z.string(),
          sign: z.union([z.literal(1), z.literal(-1)]),
          signed: z.boolean(),
        })
      ),
    })
  ),
  requirements: z.array(z.object({ id: z.string(), label: z.string() })),
})
export type IndirectCatalog = z.infer<typeof indirectCatalog>
const reference = {
  basis: z.string(),
  source_file_id: z.string().uuid().nullable(),
  source_location: z.string(),
}
const entry = z.object({ amount_minor: minor.nullable(), ...reference })
const requirement = z.object({
  status: z.enum(["reviewed", "unresolved"]),
  ...reference,
})
export const indirectRequest = z.object({
  method: methodId,
  currency: z.string().regex(/^[A-Z]{3}$/),
  start_date: z.string(),
  end_date: z.string(),
  subject: z.string(),
  entries: z.record(z.string(), entry),
  requirements: z.record(z.string(), requirement),
})
export type IndirectRequest = z.infer<typeof indirectRequest>
export const indirectEnvelope = z.object({
  case_id: z.string().uuid(),
  applied: z.literal(false),
  scenario_json: z.string(),
  scenario_sha256: z.string().regex(/^[a-f0-9]{64}$/),
  scenario_byte_count: z.number().int().positive(),
})
const indirectValue = z.object({
  schema: z.literal("loupe.financial.indirect_review/1"),
  case_id: z.string().uuid(),
  applied: z.literal(false),
  reference: z.string().url(),
  reference_section: z.string(),
  inputs: indirectRequest,
  method_label: z.string(),
  lines: z.array(
    entry.extend({
      id: z.string(),
      label: z.string(),
      sign: z.union([z.literal(1), z.literal(-1)]),
    })
  ),
  sources: z.array(
    z.object({
      id: z.string().uuid(),
      case_id: z.string().uuid(),
      filename: z.string(),
      sha256: z.string().nullable(),
    })
  ),
  missing: z.array(
    z.object({
      kind: z.enum(["amount", "review"]),
      id: z.string(),
      label: z.string(),
    })
  ),
  review_fields_complete: z.boolean(),
  difference_minor: minor.nullable(),
  limitation: z.string(),
})
const canonical = (v: unknown): string =>
  Array.isArray(v)
    ? `[${v.map(canonical).join(",")}]`
    : v && typeof v === "object"
      ? `{${Object.entries(v)
          .sort(([a], [b]) => a.localeCompare(b))
          .map(([k, x]) => `${JSON.stringify(k)}:${canonical(x)}`)
          .join(",")}}`
      : JSON.stringify(v)
export async function verifyIndirectReview(
  raw: unknown,
  catalog: IndirectCatalog,
  request: IndirectRequest
) {
  const envelope = indirectEnvelope.parse(raw),
    bytes = new TextEncoder().encode(envelope.scenario_json)
  const digest = Array.from(
    new Uint8Array(await crypto.subtle.digest("SHA-256", bytes)),
    (b) => b.toString(16).padStart(2, "0")
  ).join("")
  if (
    envelope.case_id !== catalog.case_id ||
    bytes.length !== envelope.scenario_byte_count ||
    digest !== envelope.scenario_sha256
  )
    throw Error("Workpaper capture failed its integrity check.")
  const value = indirectValue.parse(JSON.parse(envelope.scenario_json)),
    method = catalog.methods.find((m) => m.id === request.method)
  if (
    !method ||
    value.case_id !== catalog.case_id ||
    canonical(value.inputs) !== canonical(request) ||
    value.method_label !== method.label ||
    value.reference !== catalog.reference ||
    value.reference_section !== method.reference_section
  )
    throw Error("Workpaper differs from the requested scope.")
  const sourceIds = [
    ...new Set(
      [
        ...Object.values(request.entries),
        ...Object.values(request.requirements),
      ].flatMap((r) => (r.source_file_id ? [r.source_file_id] : []))
    ),
  ].sort()
  if (
    value.sources.some((s) => s.case_id !== catalog.case_id) ||
    canonical(value.sources.map((s) => s.id).sort()) !== canonical(sourceIds)
  )
    throw Error("Workpaper source scope differs.")
  const missing: { kind: string; id: string; label: string }[] = []
  let total = 0n
  if (value.lines.length !== method.terms.length)
    throw Error("Workpaper omitted an amount field.")
  for (const [i, term] of method.terms.entries()) {
    const line = value.lines[i],
      input = request.entries[term.id]
    if (
      !input ||
      canonical(line) !==
        canonical({ ...input, id: term.id, label: term.label, sign: term.sign })
    )
      throw Error("Workpaper amount field differs.")
    if (input.amount_minor !== null) {
      if (!term.signed && BigInt(input.amount_minor) < 0n)
        throw Error("Invalid amount sign.")
      total += BigInt(term.sign) * BigInt(input.amount_minor)
    }
    if (
      input.amount_minor === null ||
      !input.basis.trim() ||
      !input.source_file_id ||
      !input.source_location.trim()
    )
      missing.push({ kind: "amount", id: term.id, label: term.label })
  }
  for (const check of catalog.requirements) {
    const input = request.requirements[check.id]
    if (
      !input ||
      input.status !== "reviewed" ||
      !input.basis.trim() ||
      !input.source_file_id ||
      !input.source_location.trim()
    )
      missing.push({ kind: "review", id: check.id, label: check.label })
  }
  if (
    canonical(value.missing) !== canonical(missing) ||
    value.review_fields_complete !== (missing.length === 0) ||
    value.difference_minor !== (missing.length ? null : String(total))
  )
    throw Error("Workpaper completeness or arithmetic differs.")
  return { envelope, value }
}
export type VerifiedIndirectReview = Awaited<
  ReturnType<typeof verifyIndirectReview>
>

export function indirectMinor(
  value: string,
  currency: string,
  signed: boolean
): string | null {
  const text = value.trim(),
    negative = text.startsWith("-")
  if (negative && !signed) return null
  const amount = correctionMinor(negative ? text.slice(1) : text, currency)
  return amount === null
    ? null
    : negative && amount !== "0"
      ? `-${amount}`
      : amount
}

File: backend/services/financial/indirect_review.py
"""Guided, source-referenced indirect workpapers; never a tax or guilt finding."""
import hashlib
import json
from datetime import date
from typing import Literal
from uuid import UUID
from pydantic import BaseModel, ConfigDict, Field
from services.financial.ledger_summary import LedgerSummaryError
from services.financial.money import get_currency

REFERENCE = 'https://www.irs.gov/irm/part9/irm_09-005-009'
# Signs describe only the selected worksheet arithmetic, not source classification.
METHODS = {
    'net_worth': dict(label='Net worth', reference_section='9.5.9.5.4', terms=[
        ('closing_assets','Closing assets at cost or applicable basis',1,False),
        ('closing_liabilities','Closing liabilities and accumulated depreciation',-1,False),
        ('opening_assets','Opening assets at cost or applicable basis',-1,False),
        ('opening_liabilities','Opening liabilities and accumulated depreciation',1,False),
        ('personal_outlays','Personal expenditure and nondeductible losses',1,False),
        ('nontaxable','Adjustments for non-taxable items',-1,False),
        ('deductions','Applicable deductions and exemptions',-1,False),
        ('reported_income','Reported taxable income',-1,True)]),
    'bank_deposits': dict(label='Bank deposits',reference_section='9.5.9.7',terms=[
        ('deposits','Total deposits across the reviewed accounts',1,False),
        ('currency_expenditure','Currency expenditure outside those deposits',1,False),
        ('cash_change','Increase or decrease in cash on hand',1,True),
        ('nonincome','Non-income deposits and items, including transfers',-1,False),
        ('cost_of_goods','Cost of goods sold',-1,False),
        ('business_expenses','Business and rental expenses',-1,False),
        ('income_adjustments','Adjustments to income',-1,False),
        ('deductions','Applicable personal deductions and exemptions',-1,False),
        ('reported_income','Reported taxable income',-1,True)]),
    'expenditure': dict(label='Expenditure',reference_section='9.5.9.6.3',terms=[
        ('money_applied','Money spent or applied, including asset/liability changes',1,False),
        ('nontaxable','Non-taxable funding sources',-1,False),
        ('deductions','Applicable deductions and exemptions',-1,False),
        ('reported_income','Reported taxable income',-1,True)]),
    'cash_t': dict(label='Cash-T',reference_section='9.5.9.8.4',terms=[
        ('cash_uses','All reviewed cash uses',1,False),
        ('cash_sources','Known cash sources, including opening cash and non-taxable sources',-1,False)]),
}
REQUIREMENTS = {
    'starting_position':'Establish the starting assets, liabilities and cash on hand; explain any converted assets.',
    'nonincome_sources':'Investigate loans, gifts, inheritance, transfers, capital returns and other non-income sources.',
    'leads':'Record follow-up of reasonable explanations, alternative funding sources and unresolved leads.',
    'applicability':'Explain the period, subject, accounting basis and applicable rules, including deductions; establish the income-source basis.',
}


def indirect_methods():
    return dict(reference=REFERENCE,methods=[dict(id=key,label=value['label'],reference_section=value['reference_section'],
        terms=[dict(id=i,label=label,sign=sign,signed=signed) for i,label,sign,signed in value['terms']]) for key,value in METHODS.items()],
        requirements=[dict(id=key,label=value) for key,value in REQUIREMENTS.items()])


class IndirectEntry(BaseModel):
    model_config=ConfigDict(extra='forbid')
    amount_minor: str | None = Field(default=None,pattern=r'^(0|-?[1-9][0-9]{0,18})$')
    basis: str = Field(default='',max_length=4096)
    source_file_id: UUID | None = None
    source_location: str = Field(default='',max_length=1024)


class IndirectRequirement(BaseModel):
    model_config=ConfigDict(extra='forbid')
    status: Literal['unresolved','reviewed'] = 'unresolved'
    basis: str = Field(default='',max_length=4096)
    source_file_id: UUID | None = None
    source_location: str = Field(default='',max_length=1024)


class IndirectReviewInput(BaseModel):
    model_config=ConfigDict(extra='forbid')
    method: Literal['net_worth','bank_deposits','expenditure','cash_t']
    currency: str = Field(pattern=r'^[A-Z]{3}$')
    start_date: date
    end_date: date
    subject: str = Field(min_length=1,max_length=512)
    entries: dict[str,IndirectEntry] = Field(default_factory=dict,max_length=20)
    requirements: dict[str,IndirectRequirement] = Field(default_factory=dict,max_length=4)


def evaluate_indirect_review(case_id, request, sources):
    request=IndirectReviewInput.model_validate(request)
    get_currency(request.currency)
    if request.start_date>request.end_date or not request.subject.strip():
        raise LedgerSummaryError('Choose an ordered period and identify the subject of this review.')
    method=METHODS[request.method]
    allowed={key for key,_,_,_ in method['terms']}
    if set(request.entries)-allowed or set(request.requirements)-set(REQUIREMENTS):
        raise LedgerSummaryError('The workpaper contains fields for a different method.')
    requested_ids={str(item.source_file_id) for item in [*request.entries.values(),*request.requirements.values()] if item.source_file_id}
    indexed={source['id']:source for source in sources}
    if len(indexed)!=len(sources) or set(indexed)!=requested_ids or any(source['case_id']!=str(case_id) for source in sources):
        raise LedgerSummaryError('Workpaper sources do not match the selected case files.')
    missing=[];lines=[];total=0
    for key,label,sign,signed in method['terms']:
        entry=request.entries.get(key,IndirectEntry())
        if entry.amount_minor is not None:
            amount=int(entry.amount_minor)
            if abs(amount)>9223372036854775807 or (not signed and amount<0):
                raise LedgerSummaryError('A workpaper amount is outside its supported range or sign.')
            total+=sign*amount
        if entry.amount_minor is None or not entry.basis.strip() or not entry.source_file_id or not entry.source_location.strip():
            missing.append(dict(kind='amount',id=key,label=label))
        lines.append(dict(id=key,label=label,sign=sign,**entry.model_dump(mode='json')))
    for key,label in REQUIREMENTS.items():
        item=request.requirements.get(key,IndirectRequirement())
        if item.status!='reviewed' or not item.basis.strip() or not item.source_file_id or not item.source_location.strip():
            missing.append(dict(kind='review',id=key,label=label))
    value=dict(schema='loupe.financial.indirect_review/1',case_id=str(case_id),applied=False,
        reference=REFERENCE,reference_section=method['reference_section'],inputs=request.model_dump(mode='json'),
        method_label=method['label'],lines=lines,sources=sorted(sources,key=lambda s:s['id']),missing=missing,
        review_fields_complete=not missing,difference_minor=str(total) if not missing else None,
        limitation='Conditional arithmetic from investigator-entered amounts and recorded source references. Review fields being complete does not independently verify the sources, coverage, accounting treatment, tax liability, undeclared income or wrongdoing. No source or ledger classification is changed. Unknown amounts are not zero; investigate remaining leads before drawing conclusions.')
    content=json.dumps(value,sort_keys=True,separators=(',',':'),ensure_ascii=False)
    encoded=content.encode()
    return dict(case_id=str(case_id),applied=False,scenario_json=content,scenario_sha256=hashlib.sha256(encoded).hexdigest(),scenario_byte_count=len(encoded))
